akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -683,7 +731,80 @@ class ClusterSingletonManager(
       self ! DelayedMemberRemoved(m)
   }
 
-  def gotoOldest(): State = {
+  // Try and go to oldest, taking the lease if needed
+  def tryGoToOldest(): State = {
+    // check if lease
+    lease match {
+      case None ⇒
+        goToOldest()
+      case Some(_) ⇒
+        logInfo("Trying to acquire lease before starting singleton")
+        tryAcquireLease()
+    }
+  }
+
+  when(AcquiringLease) {
+    case Event(AcquireLeaseResult(result), _) ⇒
+      logInfo("Acquire lease result {}", result)
+      if (result) {
+        goToOldest()
+      } else {
+        setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+        stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+      }
+    case Event(Terminated(ref), AcquiringLeaseData(_, Some(singleton))) if ref == singleton ⇒
+      logInfo("Singleton actor terminated. Trying to acquire lease again before re-creating.")
+      tryAcquireLease()
+    case Event(AcquireLeaseFailure(t), _) ⇒
+      log.error(t, "failed to get lease (will be retried)")
+      setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+      stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+    case Event(LeaseRetry, _) ⇒
+      tryAcquireLease()
+    case Event(OldestChanged(oldestOption), _) ⇒
+      oldestChangedReceived = true
+      logInfo("AcquiringLease observed OldestChanged: [{} -> {}]", cluster.selfAddress, oldestOption.map(_.address))
+      oldestOption match {
+        case Some(a) if a == cluster.selfUniqueAddress ⇒
+          // already oldest
+          stay
+        case Some(a) if !selfExited && removed.contains(a) ⇒
+          handOverDone(None)
```

Comment:
`OldestChanged` is handled in same way as when `Oldest` apart from this but it does the same in the end if `Some(singleton)` is not possible here. Worth extracting this case to a method?

`AcquiringLeaseData` has a `singleton: Option[ActorRef]`, is that always on its way to terminate here (after lost lease)? `handOverDone` shouldn't be done until the singleton actor has fully Terminated so is it correct to call handOverDone from here?

Reply:
Yes it only gets there by being on its way to terminated. Its wrong to go right to handOverDone when we haven't seen the termination yet, will fix

Reply:
Also makes these the same then for acquiring and oldest so pulled out that code into a method